components/api/src/main/java/com/hotels/styx/api/plugins/spi/PluginMeterRegistry.java

```diff
@@ -0,0 +1,157 @@
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.api.plugins.spi;
+
+import io.micrometer.core.annotation.Incubating;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.search.RequiredSearch;
+import io.micrometer.core.instrument.search.Search;
+import io.micrometer.core.lang.Nullable;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.ToDoubleFunction;
+
+public class PluginMeterRegistry {
+    public static final String DEFAULT_TAG_KEY = "plugin";
+    private final MeterRegistry meterRegistry;
+    private final Tag defaultTag;
+    private Tags commonTags;
+
+    public PluginMeterRegistry(MeterRegistry meterRegistry, String pluginName) {
+        this(meterRegistry, pluginName, Tags.empty());
+    }
+
+    public PluginMeterRegistry(MeterRegistry meterRegistry, String pluginName, Tags commonTags) {
+        this.meterRegistry = meterRegistry;
+        this.defaultTag = Tag.of(DEFAULT_TAG_KEY, pluginName);
+        this.commonTags = commonTags;
+    }
+
+    public MeterRegistry getMeterRegistry() {
+        return meterRegistry;
+    }
+
+    public Tags getCommonTags() {
+        return commonTags;
+    }
+
+    public void setCommonTags(Tags commonTags) {
+        this.commonTags = commonTags;
+    }
+
+    public Tag getDefaultTag() {
+        return defaultTag;
+    }
+
+    public Collection<Meter> getMeters() {
+        return Search.in(meterRegistry).tags(Tags.of(defaultTag)).meters();
+    }
+
+    public void forEachMeter(Consumer<? super Meter> consumer) {
+        getMeters().forEach(consumer);
+    }
+
+    public MeterRegistry.Config config() {
+        return meterRegistry.config();
+    }
+
+    public Search find(String name) {
+        return Search.in(meterRegistry).tags(Tags.of(defaultTag)).name(name);
+    }
+
+    public RequiredSearch get(String name) {
+        return RequiredSearch.in(meterRegistry).tags(Tags.of(defaultTag)).name(name);
+    }
+
+    public Counter counter(String name, Iterable<Tag> tags) {
+        return meterRegistry.counter(name, commonTags.and(defaultTag).and(tags));
```

Comment:
`commonTags.and(defaultTag).and(tags)` - you use this construction a lot, when `commonTags.and(defaultTag)` is always the same value. I'd suggest that you have a field `allTags` (maybe not the best name) that you update when `commonTags` changes.
Alternatively, you could store the pluginTag as a `Tags` object rather than `Tag`. The `and(..)` method has an optimisation where, if the parameter is empty, it just returns itself, so the common case of `commonTags` being empty, and `tags` being empty, results in no actual object creation/array copying if ordered as `pluginTags.and(commonTags).and(tags)`.  This also has the nice benefit, if you carry it through to the getDefaultTag() method (which would become `Tags getPluginTags()`), that styx core could decide to provide additional plugin tags by default, without changing the interface.

Reply:
Very nice catch, I will update the code to benefit from that optimization.